File: deepspeed/inference/v2/kernels/ragged_ops/blocked_flash/blocked_flash.cpp
// Copyright (c) Microsoft Corporation.
// SPDX-License-Identifier: Apache-2.0

// DeepSpeed Team

/******************************************************************************
 * Copyright (c) 2023, Tri Dao.
 ******************************************************************************/

#include <ATen/cuda/CUDAContext.h>
#include <c10/cuda/CUDAGuard.h>
#include <torch/extension.h>

#include "blocked_flash.h"
#include "flash.h"

#define CHECK_SHAPE(x, ...)                                     \
    TORCH_CHECK(x.sizes() == torch::IntArrayRef({__VA_ARGS__}), \
                #x " must have shape (" #__VA_ARGS__ ")")

void flash_attn_by_atoms(at::Tensor& out,
                         at::Tensor& q,
                         at::Tensor& k,
                         at::Tensor& v,
                         at::Tensor& attention_atoms,
                         const float softmax_scale,
                         const bool is_causal)
{
    auto dprops = at::cuda::getCurrentDeviceProperties();

    bool is_sm8x = dprops->major == 8 && dprops->minor >= 0;
    bool is_sm90 = dprops->major == 9 && dprops->minor == 0;
    TORCH_CHECK(is_sm90 || is_sm8x, "FlashAttention only supports Ampere GPUs or newer.");

    auto q_dtype = q.dtype();
    TORCH_CHECK(q_dtype == torch::kFloat16 || q_dtype == torch::kBFloat16,
                "FlashAttention only support fp16 and bf16 data type");
    if (q_dtype == torch::kBFloat16) {
        TORCH_CHECK(is_sm90 || is_sm8x, "bfloat16 is only supported on Ampere GPUs or newer");
    }
    TORCH_CHECK(k.dtype() == q_dtype, "query and key must have the same dtype");
    TORCH_CHECK(v.dtype() == q_dtype, "query and value must have the same dtype");

    TORCH_CHECK(q.is_cuda(), "Input tensor must be on CUDA device");
    TORCH_CHECK(k.is_cuda(), "Input tensor must be on CUDA device");
    TORCH_CHECK(v.is_cuda(), "Input tensor must be on CUDA device");

    TORCH_CHECK(q.stride(-1) == 1, "Input tensor must have contiguous last dimension");
    TORCH_CHECK(k.stride(-1) == 1, "Input tensor must have contiguous last dimension");
    TORCH_CHECK(v.stride(-1) == 1, "Input tensor must have contiguous last dimension");

    const int total_q = q.size(0);
    const int head_size = k.size(-1);
    const int num_heads_kv = k.size(-2);
    const int num_heads_q = q.size(-1) / head_size;

    TORCH_CHECK(head_size <= 256, "head_size must be <= 256");
    TORCH_CHECK(head_size % 8 == 0, "head_size must be divisible by 8");
    TORCH_CHECK(num_heads_q % num_heads_kv == 0, "num_heads_q must be divisible by num_heads_kv");

    Flash_fwd_params params;

    params.is_bf16 = q.dtype() == torch::kBFloat16;

    // Set the pointers and strides.
    params.q_ptr = q.data_ptr();
    params.k_ptr = k.data_ptr();
    params.v_ptr = v.data_ptr();
    params.o_ptr = out.data_ptr();
    params.atoms = reinterpret_cast<AttentionAtom*>(attention_atoms.data_ptr());

    // All stride are in elements, not bytes.
    params.q_row_stride = q.stride(0);
    params.k_row_stride = k.stride(1);
    params.v_row_stride = v.stride(1);
    params.o_row_stride = out.stride(0);

    // Assume heads are contiguous.
    params.q_head_stride = head_size;
    params.k_head_stride = head_size;
    params.v_head_stride = head_size;
    params.o_head_stride = head_size;

    // Head params
    params.h = num_heads_q;
    params.h_k = num_heads_kv;
    params.h_h_k_ratio = num_heads_q / num_heads_kv;
    params.d = head_size;
    auto round_multiple = [](int x, int m) { return (x + m - 1) / m * m; };
    params.d_rounded = round_multiple(head_size, 32);
    params.num_atoms = attention_atoms.size(0);

    // Set the different scale values.
    params.scale_softmax = softmax_scale;
    params.scale_softmax_log2 = softmax_scale * M_LOG2E;

    params.is_causal = is_causal;

    auto stream = at::cuda::getCurrentCUDAStream().stream();
    run_mha_fwd(params, stream);
}

File: deepspeed/inference/v2/kernels/ragged_ops/blocked_flash/blocked_flash.py
# Copyright (c) Microsoft Corporation.
# SPDX-License-Identifier: Apache-2.0

# DeepSpeed Team

import torch

from deepspeed.accelerator import get_accelerator
from ....inference_utils import DtypeEnum
from deepspeed.ops.op_builder import RaggedOpsBuilder

from ... import DSKernelBase


def get_q_block_size(head_size: int) -> int:
    """
    Returns the query block size required by the kernel given a head size.
    """
    cc_major, cc_minor = torch.cuda.get_device_capability(get_accelerator().current_device())  #ignore-cuda

    if cc_major < 8:
        raise RuntimeError("Blocked attention requires CUDA compute capability >= 8.0")

    if head_size <= 64:
        return 128
    elif head_size <= 160:
        if cc_minor != 0:
            return 64
        else:
            return 128
    elif head_size == 192:
        return 128
    elif head_size == 224:
        if cc_minor != 0:
            return 64
        else:
            return 128
    else:
        if cc_major == 8 and cc_minor == 0:
            return 128
        else:
            return 64


def get_kv_block_size(head_size: int) -> int:
    """
    Return preferred granulatity for blocked KV-cache implementation.
    """
    cc_major, cc_minor = torch.cuda.get_device_capability(get_accelerator().current_device())  #ignore-cuda

    if cc_major < 8:
        raise RuntimeError("Blocked attention requires CUDA compute capability >= 8.0")

    if head_size <= 64:
        return 128
    elif head_size != 160 or cc_minor != 0:
        return 64
    else:
        return 32


class BlockedFlashAttn(DSKernelBase):
    """
    Modified implementation of flash-attn-2 tuned for inference on blocked KV-cache and wider
    range of input sequence lengths.
    """

    supported_dtypes = [DtypeEnum.fp16, DtypeEnum.bf16]

    def __init__(self, head_size: int, dtype: DtypeEnum) -> None:
        """
        Triggers any compilation of the kernels.
        """
        if not isinstance(dtype, DtypeEnum):
            dtype = DtypeEnum(dtype)

        if dtype not in BlockedFlashAttn.supported_dtypes:
            raise ValueError("Unsupported data type: {}, supported data types are {}".format(
                dtype, BlockedFlashAttn.supported_dtypes))

        # For testing, need to revert to 32
        if head_size % 16 != 0:
            raise ValueError("Head size must be divisible by 32 (configured with {})".format(head_size))

        inf_module = RaggedOpsBuilder().load()
        self.kernel = inf_module.flash_attn_by_atoms

    def __call__(self, out: torch.Tensor, q: torch.Tensor, k: torch.Tensor, v: torch.Tensor, atoms: torch.Tensor,
                 softmax_scale: float) -> torch.Tensor:
        """
        Flash attention implementation atop a blocked KV-cache. Atoms should be pre-populated.
        See attention_atom.h for further details on the structure of the information.

        Arguments:
            out (torch.Tensor): Output tensor of shape [tokens, hidden_size]
            q (torch.Tensor): Query tensor of shape [tokens, hidden_size]
            k (torch.Tensor): Key cache tensor of shape [n_blocks, block_size, n_heads_kv, head_size]. This Tensor only needs to be contiguous on the final dimension.
            v (torch.Tensor): Value cache tensor of shape [n_blocks, block_size, n_heads_kv, head_size]. This Tensor only needs to be contiguous on the final dimension.
            atoms (torch.Tensor): Atom information tensor of shape [num_atoms, 8] and type int32.
                Not all data is readable in this format. See attention_atom.h for further details.
            softmax_scale (float): Softmax scale factor.

        Returns:
            out (torch.Tensor): Output tensor of shape [tokens, hidden_size]
        """
        self.kernel(out, q, k, v, atoms, softmax_scale, True)
        return out
